mypy/join.py

```diff
@@ -57,12 +59,12 @@ def join_types(s: Type, t: Type) -> Type:
     if isinstance(s, AnyType):
         return s
 
-    if isinstance(s, NoneTyp) and not isinstance(t, Void):
-        return t
-
```

Comment:
I didn't see how; the TypeVisitor only dispatches on the type of `t`, so if `t` is an `Instance`, say, we will end up in `visit_instance`, and there's no special treatment for `UninhabitedType` there that will produce `t` as the result.